### PR TITLE
HTTP server monitor

### PR DESCRIPTION
**Pull request by jukkar**
This PR adds a way to debug incoming HTTP or HTTPS requests in the server side.

### CODE REVIEW

File: subsys/net/ip/net_shell.c [hunk 1]
```diff
@@ -928,6 +940,89 @@ int net_shell_cmd_dns(int argc, char *argv[])
...
+
+	if (*count == 0) {
+		printk("      HTTP ctx    Local           \t"
+		       "Remote          \tURL\n");
+	}
+
+	(*count)++;
+
+	printk("%s[%2d]%c %p  %16s\t%16s\t%s\n",
+	       !entry->enabled ? "-" : "", *count,
```
Comment (lpereira) [inline subsys/net/ip/net_shell.c]: Minor preference, but I'd prefer to have less cryptic output: "e" or "E" for "enabled", and "s" or "S" for "https" (maybe use Unicode if you're feeling fancy: 🔓 and 🔒, respectively U+1F513 and U+1F512).

**Review by lpereira — DISMISSED** [review]

File: subsys/net/ip/net_shell.c (hunk 1 above)
Comment (jukkar) [inline subsys/net/ip/net_shell.c]: Sure, makes sense. I will send a new version.

Comment (jukkar) [inline subsys/net/ip/net_shell.c]: BTW, I did not print the Unicode lock symbol as that might not be printed properly always.

**Comment by pfalcon** [comment]
First of all, I should say that I'm big proponent of diagnostics means in a code, so this patch is definitely good. But the baseline would be a set of compile-time options to enable request-response logging (a-la Apache and beyond).

Adding something to net shell is a bit perplexing. I'm not sure where that leads us. For example, other protocols - MQTT, COAP, etc. - should they have similar commands in the general net shell? That would mean pretty tight coupling between net shell and (optional) libraries.

One solution might to let them add separate submodules to shell, without dumping everything at the net shell. That would be more resembling of "conventional" OS, where new server (like HTTP) would add "separate commands" available via the shell, not "modify shell itself".

In the context of Zephyr those are of course subtle distinctions, but let me communication my perplexion by giving this +0 (knowing that it will be merged).

**Comment by pfalcon** [comment]
Also, I'm not sure I understand commands in "net: shell: Add HTTP server connection monitoring". So, "http" (assuming "net" shell submodule is already selected) will: a) print current HTTP connections; b) if HTTP request logging (here called "monitoring") was previously enabled, will disable it? Isn't that a bit of a single command overload?

**Comment by jukkar** [comment]
> First of all, I should say that I'm big proponent of diagnostics means in a code, so this patch is definitely good. But the baseline would be a set of compile-time options to enable request-response logging (a-la Apache and beyond).

We have already HTTP debugging option that the end user can enable, it will print everything related to requests etc. The amount of data it prints is sometimes just too much. That is why I added this monitoring support that can be activated manually via shell if that info is needed.

**Comment by jukkar** [comment]
> Adding something to net shell is a bit perplexing. I'm not sure where that leads us. For example, other protocols - MQTT, COAP, etc. - should they have similar commands in the general net shell? That would mean pretty tight coupling between net shell and (optional) libraries.

If we find out that that MQTT, CoAP etc., could benefit for some net-shell debugging aids, I see no issues here.
I am not sure what you mean by tight coupling between net-shell and libraries, if the library is not enabled, the the command would be no-op. What kind of issues you see here?

**Comment by jukkar** [comment]
> One solution might to let them add separate submodules to shell, without dumping everything at the net shell. That would be more resembling of "conventional" OS, where new server (like HTTP) would add "separate commands" available via the shell, not "modify shell itself".

Sounds quite complex to me for this simple support. But if you have good ideas for this, just send some patches.

**Comment by jukkar** [comment]
> In the context of Zephyr those are of course subtle distinctions, but let me communication my perplexion by giving this +0 (knowing that it will be merged).

This support is by no means mandatory, if people have objections to it we can easily discard this if needed.

**Comment by jukkar** [comment]
> Also, I'm not sure I understand commands in "net: shell: Add HTTP server connection monitoring". So, "http" (assuming "net" shell submodule is already selected) will: a) print current HTTP connections; b) if HTTP request logging (here called "monitoring") was previously enabled, will disable it? Isn't that a bit of a single command overload?

There are actually two commands here:

net http
- This would print currently active HTTP request(s). But if the request is short lived, then normally this command would not print anything because the request was already served.

net http monitor
- This would activate monitoring, so when a new request is served, then the type of the request (HTTPS or not), source and destination addresses and URL would be printed to output. The shell will be usable during the monitoring i.e., the input is not blocked. In order to turn off monitoring we could have "net http monitor off" or similar but that seemed an overkill so I just made it that plain "net http" will deactivate it.

**Comment by pfalcon** [comment]
> I am not sure what you mean by tight coupling between net-shell and libraries, if the library is not enabled, the the command would be no-op. What kind of issues you see here?

Well, just have a look at file path `subsys/net/ip/net_shell.c`. So, something in net/ip/ now needs to know about HTTP server library. Another way to look at that is that standard Linux netstat suddenly started to show Apache's sessions (could happen in the era of systemd's?)

Again, I don't say that there's something to change right away, just that I'm not sure how that would scale further. (E.g., having written the above, I may imagine it could be a good idea to move net_shell under subsys/net/lib/ in the future).

**Comment by pfalcon** [comment]
> In order to turn off monitoring we could have "net http monitor off" or similar but that seemed an overkill so I just made it that plain "net http" will deactivate it.

That's what I thought, thanks for confirming the idea behind doing it like that.

**Review by pfalcon — APPROVED** [review]

**Comment by jukkar** [comment]
> E.g., having written the above, I may imagine it could be a good idea to move net_shell under subsys/net/lib/ in the future

That is actually a good idea, the net-shell is not really part of IP stack actually so it does not need to be in ip directory.

**Review by tbursztyka — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]

**Review by tbursztyka — DISMISSED** [review]

**Comment by jukkar** [comment]
Uploaded fixed version because of merge conflict.

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**